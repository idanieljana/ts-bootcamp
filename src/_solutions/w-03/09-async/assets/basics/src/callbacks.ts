import * as fs from 'fs';
import * as path from 'path';

export function getInstalledAmbientTypesList(configName: string, cb: Callback<PackageJson[]>): void {
  readConfig(configName, (readError, file) => {
    if (readError) {
      logError(readError);
      cb(readError);
    } else {
      parseConfig(file!, (parseError, config) => {
        if (parseError) {
          logError(parseError);
          cb(parseError);
        } else {
          readTypesDirs(config!.dirs, (readTypesErr, typesPaths) => {
            if (readTypesErr) {
              logError(readTypesErr);
              cb(readTypesErr);
            } else {
              readPackageJsons(typesPaths!, (errPackageJsonRead, jsonsParsed) => {
                if (errPackageJsonRead) {
                  logError(errPackageJsonRead);
                  cb(errPackageJsonRead);
                } else {
                  cb(null, jsonsParsed!.map((f) => ({
                    name: f.name,
                    version: f.version,
                    description: f.description,
                  })).sort((a, b) => a.name.localeCompare(b.name)));
                }
              });
            }
          });
        }
      });
    }
  });
}

type Callback<T> = (err: Error | null, result?: T) => void;

interface Config {
  dirs: string[];
}

function isConfig(obj: any): obj is Config {
  return typeof obj === 'object'
        && 'dirs' in obj
        && Array.isArray(obj.dirs)
        && obj.dirs.every((s: any) => typeof s === 'string');
}

function parseConfig(json: string, cb: Callback<Config>): void {
  try {
    const parsed = JSON.parse(json);
    if (isConfig(parsed)) {
      cb(null, parsed);
    } else {
      cb(new Error('Error parsing config'));
    }
  } catch (error) {
    cb(new Error(`Error parsing config ${error}`));
  }
}

function readConfig(name: string, cb: Callback<string>): void {
  fs.readFile(path.join(__dirname, name), { encoding: 'utf-8' }, cb);
}

function logError(readError: Error): void {
  if (readError) {
    console.log(readError);
  }
}

function getAllTypesModules(dir: string, cb: Callback<string[]>): void {
  fs.readdir(path.join(process.cwd(), dir), cb);
}

interface TypesPath {
  name: string;
  packageJsonPath: string;
}

function readTypesDirs(dirs: string[], cb: Callback<TypesPath[]>) {
  let counter = 0;
  const resultDirs: TypesPath[] = [];
  const cwd = process.cwd();
  dirs.forEach((dir) => {
    getAllTypesModules(dir, (readDirErr, readDirs) => {
      counter++;
      if (readDirErr) {
        return cb(readDirErr);
      }
      const paths: TypesPath[] = readDirs!.map((name) => ({
        name,
        packageJsonPath: path.join(cwd, dir, name, 'package.json'),
      }));
      resultDirs.push(...paths);

      if (counter === dirs.length) {
        cb(null, resultDirs);
      }
    });
  });
}

interface PackageJson {
  name: string;
  version: string;
  description: string;
}

function readPackageJsons(paths: TypesPath[], cb: Callback<PackageJson[]>) {
  let counter = 0;
  const resultDirs: PackageJson[] = [];
  paths.forEach((filePath) => {
    fs.readFile(filePath.packageJsonPath, { encoding: 'utf-8' }, (readPackageJsonErr, file) => {
      counter++;
      if (readPackageJsonErr) {
        return cb(readPackageJsonErr);
      }
      try {
        resultDirs.push(JSON.parse(file));
      } catch (err) {
        console.log(err);
        cb(new Error(`Error parsing json file ${filePath.name} at ${filePath.packageJsonPath}`));
      }

      if (counter === paths.length) {
        cb(null, resultDirs);
      }
    });
  });
}

function isPackageJsonConfig(obj: any): obj is PackageJson {
  return typeof obj === 'object'
        && 'description' in obj
        && typeof obj.description === 'string';
}

export function parsePackageJsonDescription(packageJsonPath: string, cb: Callback<string>): void {
  fs.readFile(packageJsonPath, { encoding: 'utf-8' }, (err, file) => {
    if (err) {
      cb(err);
    } else {
      try {
        const parsed = JSON.parse(file);
        if (isPackageJsonConfig(parsed)) {
          cb(null, parsed.description);
        } else {
          cb(new Error('Incorrect config provided or no description in the file'));
        }
      } catch (err) {
        cb(err);
      }
    }
  });
}
