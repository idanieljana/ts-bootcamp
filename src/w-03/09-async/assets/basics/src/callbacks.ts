import * as fs from "fs";
import * as path from "path";

export function getInstalledAmbientTypesList(configName: string, cb: Callback<PackageJson[]>): void {
    readConfig(configName, (readError, file) => {
        if (readError) {
            logError(readError);
            cb(readError)
        } else {
            parseConfig(file!, (parseError, config) => {
                if (parseError) {
                    logError(parseError);
                    cb(parseError)
                } else {
                    readTypesDirs(config!.dirs, (readTypesErr, typesPaths) => {
                        if (readTypesErr) {
                            logError(readTypesErr);
                            cb(readTypesErr)
                        } else {
                            readPackageJsons(typesPaths!, (errPackageJsonRead, jsonsParsed) => {
                                if (errPackageJsonRead) {
                                    logError(errPackageJsonRead);
                                    cb(errPackageJsonRead)
                                } else {
                                    cb(null, jsonsParsed!.map(f => {
                                        return {
                                            name: f.name,
                                            version: f.version,
                                        }
                                    }).sort((a, b) => {
                                        return a.name.localeCompare(b.name)
                                    }));
                                }
                            })

                        }
                    })
                }
            })
        }
    })
}

type Callback<T> = (err: Error | null, result?: T) => void;

interface Config {
    dirs: string[];
}

function isConfig(obj: any): obj is Config {
    return typeof obj === "object"
        && "dirs" in obj
        && Array.isArray(obj.dirs)
        && obj.dirs.every((s: any) => typeof s === "string")
}

function parseConfig(json: string, cb: Callback<Config>): void  {
    try {
        const parsed = JSON.parse(json);
        if (isConfig(parsed)) {
            cb(null, parsed)
        } else {
            cb(new Error(`Error parsing config`))
        }
    } catch (error) {
        cb(new Error(`Error parsing config ${error}`))
    }
}

function readConfig(name: string, cb: Callback<string>): void {
    fs.readFile(path.join(__dirname, name), { encoding: "utf-8" }, cb);
}

function logError(readError: Error): void {
    if (readError) {
        console.log(readError);
    }
}

function getAllTypesModules(dir: string, cb: Callback<string[]>): void {
    fs.readdir(path.join(process.cwd(), dir), cb)
}

interface TypesPath {
    name: string;
    packageJsonPath: string;
}

function readTypesDirs(dirs: string[], cb: Callback<TypesPath[]>) {
    let counter = 0;
    const resultDirs: TypesPath[] = []
    const cwd = process.cwd();
    dirs.forEach(dir => {
        getAllTypesModules(dir, (readDirErr, readDirs) => {
            counter++;
            if (readDirErr) {
                return cb(readDirErr)
            } else {
                const paths: TypesPath[] = readDirs!.map(name => {
                    return {
                        name: name,
                        packageJsonPath: path.join(cwd, dir, name, "package.json"),
                    };
                })
                resultDirs.push(...paths)
            }
            if (counter === dirs.length) {
                cb(null, resultDirs)
            }
        })
    })
}

interface PackageJson {
    name: string;
    version: string;
}

function readPackageJsons(paths: TypesPath[], cb: Callback<PackageJson[]>) {
    let counter = 0;
    const resultDirs: PackageJson[] = []
    paths.forEach(filePath => {
        fs.readFile(filePath.packageJsonPath, { encoding: "utf-8" }, (readPackageJsonErr, file) => {
            counter++;
            if (readPackageJsonErr) {
                return cb(readPackageJsonErr)
            } else {
                try {
                    resultDirs.push(JSON.parse(file))
                } catch(err) {
                    console.log(err)
                    cb(new Error(`Error parsing json file ${filePath.name} at ${filePath.packageJsonPath}`))
                }
            }
            if (counter === paths.length) {
                cb(null, resultDirs)
            }
        });
    })
}
