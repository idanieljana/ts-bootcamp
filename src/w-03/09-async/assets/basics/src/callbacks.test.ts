import { getInstalledAmbientTypesList } from "./callbacks";

/**
 * Note: promisify is a Node.js core utility which could be used to
 * improve experience with Node.js callback APIs
 */
import { promisify } from "util";

describe("AmbientTypesParser", () => {
    test("should parse config and return installed ambient types", async () => {
        /**
         * Note: npm module `mock-fs` should be used for unit testing with fs
         * https://github.com/tschaub/mock-fs
         */
        const path = "./types_config.json";
        const result = await promisify(getInstalledAmbientTypesList)(path);
        expect(result).toMatchSnapshot();
    });
})