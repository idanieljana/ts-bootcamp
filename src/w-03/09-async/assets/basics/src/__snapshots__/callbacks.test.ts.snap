// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmbientTypesParser should parse config and return installed ambient types 1`] = `
Array [
  Object {
    "name": "@types/babel__core",
    "version": "7.1.10",
  },
  Object {
    "name": "@types/babel__generator",
    "version": "7.6.2",
  },
  Object {
    "name": "@types/babel__template",
    "version": "7.0.3",
  },
  Object {
    "name": "@types/babel__traverse",
    "version": "7.0.15",
  },
  Object {
    "name": "@types/graceful-fs",
    "version": "4.1.3",
  },
  Object {
    "name": "@types/istanbul-lib-coverage",
    "version": "2.0.3",
  },
  Object {
    "name": "@types/istanbul-lib-report",
    "version": "3.0.0",
  },
  Object {
    "name": "@types/istanbul-reports",
    "version": "3.0.0",
  },
  Object {
    "name": "@types/jest",
    "version": "26.0.14",
  },
  Object {
    "name": "@types/json-schema",
    "version": "7.0.6",
  },
  Object {
    "name": "@types/json5",
    "version": "0.0.29",
  },
  Object {
    "name": "@types/node",
    "version": "14.11.2",
  },
  Object {
    "name": "@types/normalize-package-data",
    "version": "2.4.0",
  },
  Object {
    "name": "@types/prettier",
    "version": "2.1.1",
  },
  Object {
    "name": "@types/stack-utils",
    "version": "1.0.1",
  },
  Object {
    "name": "@types/yargs",
    "version": "15.0.7",
  },
  Object {
    "name": "@types/yargs-parser",
    "version": "15.0.0",
  },
]
`;
